File: leetcode/company-wise/uber/0269_alien_dictionary.py
# LC 269. Alien Dictionary | Hard | Uber, Google
from collections import defaultdict, deque

class Solution:
    def alienOrder(self, words):
        adj = defaultdict(set)
        indeg = {c: 0 for w in words for c in w}
        for i in range(len(words)-1):
            w1, w2 = words[i], words[i+1]
            minLen = min(len(w1), len(w2))
            if len(w1) > len(w2) and w1[:minLen] == w2[:minLen]: return ""
            for j in range(minLen):
                if w1[j] != w2[j]:
                    if w2[j] not in adj[w1[j]]:
                        adj[w1[j]].add(w2[j]); indeg[w2[j]] += 1
                    break
        q = deque(c for c in indeg if indeg[c] == 0)
        res = []
        while q:
            c = q.popleft(); res.append(c)
            for nei in adj[c]:
                indeg[nei] -= 1
                if indeg[nei] == 0: q.append(nei)
        return ''.join(res) if len(res) == len(indeg) else ""

File: leetcode/company-wise/uber/0269_alien_dictionary.cpp
// LC 269. Alien Dictionary | Hard | Uber, Google
#include <string>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
public:
    string alienOrder(vector<string>& words) {
        unordered_map<char, unordered_set<char>> adj;
        unordered_map<char, int> indeg;
        for (auto& w : words) for (char c : w) indeg[c] = 0;
        for (int i = 0; i < (int)words.size()-1; i++) {
            string &w1 = words[i], &w2 = words[i+1];
            int mn = min(w1.size(), w2.size());
            if (w1.size() > w2.size() && w1.substr(0,mn) == w2.substr(0,mn)) return "";
            for (int j = 0; j < mn; j++)
                if (w1[j] != w2[j]) { if (!adj[w1[j]].count(w2[j])) { adj[w1[j]].insert(w2[j]); indeg[w2[j]]++; } break; }
        }
        queue<char> q; for (auto& [c, d] : indeg) if (d == 0) q.push(c);
        string res;
        while (!q.empty()) { char c = q.front(); q.pop(); res += c; for (char n : adj[c]) if (--indeg[n]==0) q.push(n); }
        return res.size() == indeg.size() ? res : "";
    }
};
